### PR TITLE
Linting, refactoring and other improvements

### PR DESCRIPTION
**Pull request by teutoburg**
Listen to Pylint, also docstring formatting and refining, some improvements to astropy units usage (less `ìsinstance`), remove some duplicate checks and conversions.

Some general refactoring like extracting duplicate code parts (of which there were a bunch!), de-nesting, using generator, et cetera.

Improving the "add function call to meta" decorator and functions, which now actually works I think.

Fixing a bug reported by a user about redshift and magnitude scaling not passed in the `galaxies.galaxy` function.

Adding some tests and updating others, fix a warning or two.

### CODE REVIEW

**Review by hugobuddel — APPROVED** [review]
Thanks, this is a bit big, but needed to be done anyway. I think I caught two mistakes; let's assume those are the only two!

Perhaps in the future we should separate stylistic changes from changes to the code (like the decorator). Now it was a bit hard to figure out why certain changes were made. 

Nevertheless, it is good to see the code cleaned up like this, thanks.

And we probably don't have to do this often, so it's fine. We need to break some eggs to make an omelet.

It would be good to run the tests and notebooks from ScopeSim and the IRDB against these changes though, but we can do that after we merge it.

File: scopesim_templates/extragalactic/exgal_models.py [hunk 1]
```diff
@@ -244,98 +247,87 @@ def intensity(self):
...
-                           q=self.q)
-            result = mod(self.x, self.y)
-        else:
-            result = np.ones(shape=self.x.shape)
+        result : numpy.ndarray
+            Velocity map.
 
-        return result
+        """
+        if self.sigma <= 0.0:
```
[the remark this answers is not in the document]
Comment (teutoburg) [reply scopesim_templates/extragalactic/exgal_models.py:260]: You're right, I did the dispersion one first and then copied it here, oops

File: scopesim_templates/extragalactic/galaxies.py [hunk 2]
```diff
@@ -305,39 +282,40 @@ def spiral_two_component(extent=60*u.arcsec, fluxes=(0, 0), offset=(0, 0)):
...
+        field.header["CTYPE2"] = "DEC--TAN"
+        field.header["SPEC_REF"] = field.header["SPEC_EXT"] - spec_ext
+
+    # TODO: scale image plane according to fluxes
+    # TODO: shift header values according to offset
+
+    # src.meta.update(params)
     # Ensure the number of _meta_dicts is the same as the number of _fields.
+    # TODO: check if this still works with the new decorator...
     src._meta_dicts += [{}] * (len(src.fields) - len(src._meta_dicts))
```
[the remark this answers is not in the document]
Comment (teutoburg) [reply scopesim_templates/extragalactic/galaxies.py:307]: I never encountered the `src._meta_dicts` attribute before tbh...

[the remark this answers is not in the document]
Comment (hugobuddel) [reply scopesim_templates/extragalactic/galaxies.py:307]: It is for composed sources. You can add two sources, and then you get one new source, but then it becomes kinda undefined what should be in `self.meta` of that combined source. So in addition to `self.meta`, there is also `self._meta_dicts`  which has all the `meta`s of the individual components. Or something like that, I'm not entirely sure I'm remembering it right. But something to keep in the back of our minds.

File: scopesim_templates/extragalactic/galaxies.py [hunk 3]
```diff
@@ -74,44 +94,45 @@ def galaxy(sed,           # The SED of the galaxy
...
-    if isinstance(pixel_scale, u.Quantity) is False:
-        pixel_scale = pixel_scale * u.arcsec
-    if isinstance(r_eff, u.Quantity) is False:
-        r_eff = r_eff * u.arcsec
+    gal, pixel_scale = _galaxy_setup(pixel_scale, r_eff, extend,
+                                     n=n, ellip=ellip, theta=theta)
+
     if isinstance(sed, str):
-        sp = Spextrum(sed).redshift(z=z)
-        scaled_sp = sp.scale_to_magnitude(amplitude=amplitude, filter_curve=filter_curve)
```
[the remark this answers is not in the document]
Comment (hugobuddel) [reply scopesim_templates/extragalactic/galaxies.py:89]: Thanks.

**Comment by hugobuddel** [comment]
Merge? (I'm not pressing that button at 21:00 on a Friday night.)

**Comment by teutoburg** [comment]
Other people pay money to watch things go up in flames on a Friday night 😛

**Pull request merged**